### PR TITLE
feat(macOS): add native elements support for macOS/Skia

### PR DESCRIPTION
**Pull request by spouliot**
GitHub Issue (If applicable): closes #

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

It is not possible to use native controls inside a Skia-based app on macOS.

## What is the new behavior?

It's now possible to integrate native controls (some code required) inside Skia-based apps on macOS.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

https://github.com/unoplatform/uno-private/issues/479

### CODE REVIEW

File: src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs [hunk 1]
```diff
@@ -80,7 +82,15 @@ private void Draw(SKSurface surface)
...
 			if (RootElement?.Visual is { } rootVisual)
 			{
-				RootElement.XamlRoot?.Compositor.RenderRootVisual(surface, rootVisual, null);
+				int width = (int)nativeWidth;
+				int height = (int)nativeHeight;
+				var path = SkiaRenderHelper.RenderRootVisualAndReturnPath(width, height, rootVisual, surface);
+				if (path is { })
+				{
+					var svg = path.ToSvgPathData();
+					// remove first path since it cover the whole area
```
Comment (spouliot) [inline src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: @ramezgerges I'm not sure why but the first path returned covers everything, which clips all native elements. I removed it but the problem seems elsewhere.

Comment (ramezgerges) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: I just checked and `RenderRootVisualAndReturnPath` does indeed add the entire window viewport first, but it's intended. The idea is that we start with nothing clipped, and then we "chip away" at this full viewport whenever we find a native visual.

Notice the line with `mainPath.AddRect(new SKRect(0, 0, width, height));`
```csharp
			// the entire viewport
			var mainPath = new SKPath();
			mainPath.AddRect(new SKRect(0, 0, width, height));

			rootVisual.Compositor.RenderRootVisual(surface, rootVisual, (session, visual) =>
			{
				// minus the native-element areas
				if (visual.IsNativeHostVisual)
				{
					_clipPath.Reset();
					_clipPath.AddRect(canvas.DeviceClipBounds);
					_visualPath.Reset();
					_visualPath.AddRect(visual.TotalMatrix.ToMatrix3x2().Transform(new Rect(new Point(), visual.Size.ToSize())).ToSKRect());
					mainPath = mainPath.Op(_clipPath.Op(_visualPath, SKPathOp.Intersect), SKPathOp.Difference);
				}

				// plus the popups
				if (visual.IsPopupVisual)
				{
					_clipPath.Reset();
					_clipPath.AddRect(canvas.DeviceClipBounds); // possible QoL improvement: find a way to get the tight clip path (e.g. for rounded corners) not just the bounding rect
					_visualPath.Reset();
					_visualPath.AddRect(visual.TotalMatrix.ToMatrix3x2().Transform(new Rect(new Point(), visual.Size.ToSize())).ToSKRect());
					// note that popups are always traversed last in the tree, so there are no "races" between the paths of PopupVisuals and NativeHostVisuals
					mainPath = mainPath.Op(_clipPath.Op(_visualPath, SKPathOp.Intersect), SKPathOp.Union);
				}
			});
```

Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: Thanks for looking! I think (from the output I get) that the SVG path (made by Skia) is skipping the `Op` 😢

Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: This is what I get from `path.ToSvgPathData`:

```
M2048 0L0 0L0 1480L2048 1480L2048 0ZM776 240L613 240L613 255L776 255L776 240Z
```

Comment (ramezgerges) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: I'm not an SVG guru, but I think this is actually accurate! The thing to watch out for is the `FillType` (aka `fill-rule` in Web APIs). I've adjusted your string a tiny bit to `M2048 0L0 0L0 1480L2048 1480L2048 0ZM776 240L613 240L613 355L776 355L776 240Z` (I made the "native viewport" slightly bigger). The FillType is `EvenOdd` by default. Here's a snippet with your string. Notice that if you remove the `fill-rule` bit, you get a solid rectangle like you experience: https://www.svgviewer.dev/s/nK750twd

File: src/Uno.UI.Runtime.Skia.MacOS/MacOSNativeElementHostingExtension.cs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+	public static void Register() => ApiExtensibility.Register<ContentPresenter>(typeof(ContentPresenter.INativeElementHostingExtension), o => new MacOSNativeElementHostingExtension(o));
+
+	public void ArrangeNativeElement(object content, Rect arrangeRect, Rect clipRect)
+	{
+		if (content is MacOSNativeElement element)
+		{
+			if (element.Detached)
+			{
+				this.Log().Debug($"Cannot arrange element `{nameof(content)}` of type {content.GetType().FullName} since it was detached.");
```
Comment (ramezgerges) [inline src/Uno.UI.Runtime.Skia.MacOS/MacOSNativeElementHostingExtension.cs:38]: Maybe also a `Debug.Assert`?

File: src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs (hunk 1 above)
Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSWindowHost.cs:91]: good catch, the default is actually `kCAFillRuleNonZero`
and I'll need to reverse the shape (which was the original code)

File: src/Uno.UI.Runtime.Skia.MacOS/MacOSMetalRenderer.cs [hunk 3]
```diff
@@ -15,7 +15,7 @@ internal static class MacOSMetalRenderer
 		// FIXME: contribute some extra API (e.g. using `nint` or `IntPtr`) to SkiaSharp to avoid reflection
 		// net8+ alternative -> https://steven-giesel.com/blogPost/05ecdd16-8dc4-490f-b1cf-780c994346a4
 		var get = typeof(GRContext).GetMethod("GetObject", BindingFlags.Static | BindingFlags.NonPublic);
```
[the remark this answers is not in the document]
Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSMetalRenderer.cs:17]: https://github.com/mono/SkiaSharp/pull/3070

File: src/Uno.UI.Runtime.Skia.MacOS/MacOSNativeElementHostingExtension.cs (hunk 2 above)
Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/MacOSNativeElementHostingExtension.cs:38]: It's not a theoretical issue as it happens while running SampleApp. Using `Debug.Assert` would be harmful for normal (app) debugging.

File: src/Uno.UI.Runtime.Skia.MacOS/UnoNativeMac/UnoNativeMac/UNOWindow.m [hunk 4]
```diff
@@ -639,6 +656,95 @@ void uno_set_window_close_callbacks(window_should_close_fn_ptr shouldClose, wind
...
+#if DEBUG
+        NSLog(@"uno_window_clip_svg %@ %@ %s", window, window.contentView.layer.description, svg);
+#endif
+        NSArray<__kindof NSView *> *subviews = window.contentViewController.view.subviews;
+        for (int i = 0; i < subviews.count; i++) {
+            NSView* view = subviews[i];
+#if DEBUG
+            NSLog(@"uno_window_clip_svg subview %d %@ layer %@ mask %@", i, view, view.layer, view.layer.mask);
+#endif
+            CGMutablePathRef path = CGPathCreateMutable();
```
Comment (jeromelaban) [inline src/Uno.UI.Runtime.Skia.MacOS/UnoNativeMac/UnoNativeMac/UNOWindow.m:687]: Would it be of interest to cache those instances based on the string hash?

Comment (spouliot) [reply src/Uno.UI.Runtime.Skia.MacOS/UnoNativeMac/UnoNativeMac/UNOWindow.m:687]: Worth measuring. There's a nice [`NSCache`](https://developer.apple.com/documentation/foundation/nscache?language=objc) available (so we don't have to do the hashing ourselves).

**Pull request merged**